### PR TITLE
Indu

### PR DESCRIPTION
**Pull request by Indu2204**
I have made changed in the code corresponding to task 2

### CODE REVIEW

File: create_cluster.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import os
+import datetime
+
+from google.colab import auth
+import gspread
+from google.auth import default  #authenticatiing to google
+from clusters.py import all
+
+def CreateClusterV1(filename: str,
+                    features: List[str],
```
Comment (MalcolmSlaney) [inline create_cluster.py]: OK, but there is no default for features.

File: create_cluster.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+from sklearn.datasets import make_blobs
+import json
+from sklearn.utils._openmp_helpers import _openmp_effective_n_threads
+import os
+import datetime
+
+from google.colab import auth
+import gspread
+from google.auth import default  #authenticatiing to google
+from clusters.py import all
```
Comment (MalcolmSlaney) [inline create_cluster.py]: I'm not convinced that moving the cluster routine to a new function is a good idea.

Most importantly, imports are weird in Python, and depending on how the package is set up, this might or might not be found.

More importantly, when you do an import all, it's not clear later where a string comes from.  Thus it is not allowed by the Google style.  Now I understand why.

File: create_cluster.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+def CreateClusterV1(filename: str,
+                    features: List[str],
+                    duplicate_column_name: str,
+                    labels: List[str] = ['R250',	'R500',	'R1000',	'R2000',	'R3000',	'R4000',	'R6000',	'R8000'],
+                    spreadsheet_path: str = spreadsheet_v1_path,
+                    save_path: str = default_cluster_dir,
+                    cluster_labels: dict = {4: 'Low flat',
```
Comment (MalcolmSlaney) [inline create_cluster.py:59]: We should really not build in this dictionary... a later version of the code that does the cluster assignments might assign each cluster a different name.

Much better to figure out which cluster belongs with each name.  Perhaps using the mean and slope of each cluster as a feature, and then do nearest-neighbor lookup to assign the name.  Make sense?

File: create_cluster.py [hunk 4]
```diff
@@ -21,8 +21,38 @@
 from clusters import *
 
 def CreateClusterV1(filename: str,
-                    features: List[str],
-                    duplicate_column_name: str,
+                    features: List[str] = [
+        "AgeAtTestDate",
```
Comment (MalcolmSlaney) [inline create_cluster.py:25]: Oh, ok.  But this list should be defined outside the function definition so that others can refer to it.  And then function default should be the previously defined name.  OK?  Same with duplicate_column_name.

**Review by MalcolmSlaney — COMMENTED** [review]
Thanks.

I am not convinced that pulling the cluster code into a new file is a good thing.  It does make the import problematic.  

Why do you think this is a good thing?  (I'm convincible.)

File: create_cluster.py (hunk 3 above)
Comment (Indu2204) [reply create_cluster.py:59]: I was thinking of resolving task by task as mentioned on the spreadsheet. Hence, I did not proceed to the step of finding the mean and slope of cluster. I am working on it now. Will commit the changes in a while

File: create_cluster.py (hunk 4 above)
Comment (Indu2204) [reply create_cluster.py:25]: Like a global variable (for both duplicate_column_name and features)? 

There was no default previously. I included default only in this version. Let me know if I should modify this.

**Comment by Indu2204** [comment]
I thought it will be easily accessible for the users. By creating a new file that has the function that does all the steps in one go, the function is more easily visible to the users. While going through the clusters.py previously, create cluster function was camouflaged. But I see the point you are making here. It could make the import problematic.  One will need access to clusters.py to utilise create cluster function.

File: create_cluster.py (hunk 4 above)
Comment (MalcolmSlaney) [reply create_cluster.py:25]: Yes, a global name for both pieces of information, like you have for the other one... make sure v1 is in the name.  Hopefully similar names.

**Comment by Indu2204** [comment]
> Thanks.
> 
> I am not convinced that pulling the cluster code into a new file is a good thing. It does make the import problematic.
> 
> Why do you think this is a good thing? (I'm convincible.)

I thought it will be easily accessible for the users. By creating a new file that has the function that does all the steps in one go, the function is more easily visible to the users. While going through the clusters.py previously, create cluster function was camouflaged. But I see the point you are making here. It could make the import problematic. One will need access to clusters.py to utilise create cluster function.

File: clusters.py [hunk 5]
```diff
@@ -7,7 +9,12 @@
 of the user.
 """
 spreadsheet_v1_path = 'https://docs.google.com/spreadsheets/d/119_-qrfzGXwV1YBUJdBzvtAQTZnl-xwN7hD9FK5SWfU/edit#gid=84023254'
+features_before = [ "AgeAtTestDate", "R250", "R500", "R1000", "R2000", "R3000", "R4000", "R6000", "R8000",
```
Comment (MalcolmSlaney) [inline clusters.py:12]: Good, but I'm not sure what this name means.  Before what?

the AC and BC fields make sense.  Are we using the other scores?

Not sure what the right answer is... perhaps features_v1?

File: clusters.py [hunk 6]
```diff
@@ -7,7 +9,12 @@
...
+              "L250", "L500", "L1000", "L2000", "L3000", "L4000", "L6000", "L8000", "RBone500",
+              "RBone1000", "RBone2000", "RBone4000", "LBone500", "LBone1000", "LBone2000", "LBone4000",
+              "MonSNR_Score_R", "Word_Rec_Score_R", "MonSNR_Score_L", "Word_Rec_Score_L" ]
 
+duplicate_column_name_1: str = 'LBone200'
```
Comment (MalcolmSlaney) [inline clusters.py]: This should be v1 not just 1.

File: clusters.py [hunk 7]
```diff
@@ -569,3 +576,118 @@ def LoadFromJson(
...
+        custom_spreadsheet_path = 'path/to/custom_spreadsheet.xlsx'
+        cleaned_custom_df = CleanData(features=custom_features,
+                                      duplicate_column_name=custom_duplicate_column,
+                                      spreadsheet_path=custom_spreadsheet_path)
+    """
+  
+  rows = ImportSpreadsheet(spreadsheet_path)
+  rows = RenameDuplicateColumns(rows, duplicate_column_name)
+  data = MakeDataClass(rows[0], rows, features)
+  df = ConvertToPanda(data, features)
```
Comment (MalcolmSlaney) [inline clusters.py:630]: All good above.  But I think I'd leave the final panda version off.. I don't use it.  :-)

**Review by MalcolmSlaney — COMMENTED** [review]
Good.  Getting closer.

File: clusters.py (hunk 5 above)
Comment (Indu2204) [reply clusters.py:12]: I used the terms features_before and features_after to distinguish the features that will have new columns after HLClassification() and Kmeans cluastering.

Now that I see, features_initial will make more sense. I will leave a comment also nearby.

File: clusters.py [hunk 8]
```diff
@@ -7,7 +9,21 @@
...
+# features_before = [ "AgeAtTestDate", "R250", "R500", "R1000", "R2000", "R3000", "R4000", "R6000", "R8000",
+#               "L250", "L500", "L1000", "L2000", "L3000", "L4000", "L6000", "L8000", "RBone500",
+#               "RBone1000", "RBone2000", "RBone4000", "LBone500", "LBone1000", "LBone2000", "LBone4000",
+#               "MonSNR_Score_R", "Word_Rec_Score_R", "MonSNR_Score_L", "Word_Rec_Score_L" ]
+
+duplicate_column_name_1: str = 'LBone2000'
 
+#Golden cluster can be computed by calling CreateKMeans() with random_seed = 0
+
+golden_cluster = {(69.764465, 17.85707): 'High flat',
```
Comment (MalcolmSlaney) [inline clusters.py:21]: Good. 

1) You should add v1 to the name.  
2) It doesn't make sense to use floats as a dictionary key.  Better to use the strings as a key.

File: clusters.py [hunk 9]
```diff
@@ -320,6 +336,23 @@ def MakeDataClass(column_names: List,
 
   return data
 
+def CheckBCandAC(data: pd.DataFrame, threshold: int = 100) -> pd.DataFrame:
+  """
+  A function to drop the rows where Bone Conduction is greater than Air 
+  Conduction by atleast 10dB
```
Comment (MalcolmSlaney) [inline clusters.py:342]: Comment is wrong, it's no longer 10dB.  You should also explain that the default, 100, means that no data is removed.

File: clusters.py [hunk 10]
```diff
@@ -320,6 +336,23 @@ def MakeDataClass(column_names: List,
...
+
+  Args:
+  - data: A pandas dataframe on which HLossClassifier() function is run.
```
Comment (MalcolmSlaney) [inline clusters.py:345]: Change HLossClassifier -> CheckBCandAC

File: clusters.py [hunk 11]
```diff
@@ -567,9 +602,105 @@ def LoadFromJson(
 
   return kmeans, features_before, cluster_labels, features_after
 
+def ReadPreprocessData(duplicate_column_name: str = duplicate_column_name_1,
+              spreadsheet_path: str = spreadsheet_v1_path) -> np.ndarray:
+
+  """
+    Clean and transform data from a spreadsheet into a pandas DataFrame.
```
Comment (MalcolmSlaney) [inline clusters.py:609]: This is no longer returning a DF.

File: clusters.py [hunk 12]
```diff
@@ -7,7 +9,21 @@
 of the user.
 """
 spreadsheet_v1_path = 'https://docs.google.com/spreadsheets/d/119_-qrfzGXwV1YBUJdBzvtAQTZnl-xwN7hD9FK5SWfU/edit#gid=84023254'
+# features_before = [ "AgeAtTestDate", "R250", "R500", "R1000", "R2000", "R3000", "R4000", "R6000", "R8000",
+#               "L250", "L500", "L1000", "L2000", "L3000", "L4000", "L6000", "L8000", "RBone500",
+#               "RBone1000", "RBone2000", "RBone4000", "LBone500", "LBone1000", "LBone2000", "LBone4000",
+#               "MonSNR_Score_R", "Word_Rec_Score_R", "MonSNR_Score_L", "Word_Rec_Score_L" ]
+
+duplicate_column_name_1: str = 'LBone2000'
```
Comment (MalcolmSlaney) [inline clusters.py:17]: Change name_1 to name_v1 (it wasn't clear what the 1 meant the first time I read this.)

File: clusters.py [hunk 13]
```diff
@@ -567,9 +602,105 @@ def LoadFromJson(
...
+        A dictionary containing the updated cluster labels for each centroid in `kmeans`. The keys represent the index of the centroid in `kmeans`,
+        and the values are the corresponding cluster labels from the `golden_cluster` dictionary.
+  """
+  
+  clusters_slope_mean = SlopeandMean(kmeans)
+  new_labels = {}
+  golden_cluster_centroids = list(golden_cluster.keys())
+  golden_cluster_centroids
+
+  for mean_slope, i in zip(clusters_slope_mean, range(len(clusters_slope_mean))):
```
Comment (MalcolmSlaney) [inline clusters.py:689]: It is better to use 
   for i, (mean, slope) in enumerate(clusters_slope_mean):
(I think this works).

File: clusters.py [hunk 14]
```diff
@@ -567,9 +602,105 @@ def LoadFromJson(
...
+          new_labels[i] = golden_cluster[golden_cluster_centroids[index]]
+
+  return new_labels
+
+
 def CreateClusterV1(filename: str,
-                    features: List[str],
-                    duplicate_column_name: str,
+                    duplicate_column_name: str = duplicate_column_name_1,
                     labels: List[str] = ['R250',	'R500',	'R1000',	'R2000',	'R3000',	'R4000',	'R6000',	'R8000'],
```
Comment (MalcolmSlaney) [inline clusters.py:704]: The labels default should also be a global variable so others can access it.

File: clusters.py (hunk 8 above)
Comment (Indu2204) [reply clusters.py:21]: Done.

**Comment by MalcolmSlaney** [comment]
Good progress... I'm merging this so I can look at the big picture... and we need to change the keys and values for the mean-slope dictionary.

Thanks.
